src/shared/codelens/csharpCodeLensProvider.ts

```diff
@@ -0,0 +1,305 @@
+ * @param lambdaLocalInvokeParams - Information about the Lambda Handler to invoke locally
+ */
+async function onLocalInvokeCommand({
+    commandName,
+    lambdaLocalInvokeParams,
+}: {
+    commandName: string,
+    lambdaLocalInvokeParams: LambdaLocalInvokeParams,
+}): Promise<{ datum: Datum }> {
+
+    const runtime = await getRuntimeForLambda({
+        handlerName: lambdaLocalInvokeParams.handlerName,
+        templatePath: lambdaLocalInvokeParams.samTemplate.fsPath
+    })
+
+    // TODO : Implement local run/debug in future backlog tasks
+    vscode.window.showInformationMessage(
+        `Local ${lambdaLocalInvokeParams.isDebug ? 'debug' : 'run'} support for csharp is currently not implemented.`
+    )
+
+    return getMetricDatum({
+        isDebug: lambdaLocalInvokeParams.isDebug,
+        command: commandName,
+        runtime,
+    })
+}
+
+export async function makeCSharpCodeLensProvider(): Promise<vscode.CodeLensProvider> {
+    const logger = getLogger()
+
+    const codeLensProvider: vscode.CodeLensProvider = {
+        provideCodeLenses: async (
+            document: vscode.TextDocument,
+            token: vscode.CancellationToken
+        ): Promise<vscode.CodeLens[]> => {
+            const handlers: LambdaHandlerCandidate[] = await getLambdaHandlerCandidates(document)
+            logger.debug(
+                'csharpCodeLensProvider.makePythonCodeLensProvider handlers:',
+                JSON.stringify(handlers, undefined, 2)
+            )
+
+            return makeCodeLenses({
+                document,
+                handlers,
+                token,
+                language: 'csharp'
+            })
+        }
+    }
+
+    return codeLensProvider
+}
+
+export async function getLambdaHandlerCandidates(document: vscode.TextDocument): Promise<LambdaHandlerCandidate[]> {
+    const assemblyName = await getAssemblyName(document.uri)
+    if (!assemblyName) {
+        return []
+    }
+
+    const symbols: vscode.DocumentSymbol[] = (
+        (await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
+            'vscode.executeDocumentSymbolProvider',
+            document.uri
+        )) || []
+    )
+
+    return getLambdaHandlerSymbolsTuplets(document, symbols)
+        .map<LambdaHandlerCandidate>(tuplet => {
+            const handlerName = produceHandlerName(assemblyName, tuplet)
+
+            return {
+                filename: document.uri.fsPath,
+                handlerName,
+                range: tuplet.method.range,
+            }
+        })
+}
+
+async function getAssemblyName(sourceCodeUri: vscode.Uri): Promise<string | undefined> {
+    const projectFile: vscode.Uri | undefined = await findParentProjectFile(sourceCodeUri)
+
+    if (!projectFile) {
+        return undefined
+    }
+
+    // TODO : Perform an XPATH parse on the project file
+    // If Project/PropertyGroup/AssemblyName exists, use that. Otherwise use the file name.
```

Comment:
What if the `AssemblyName` is specified in a `.targets`/`.props` file, and does not match the filename?